immunopepper/traversal.py

```diff
@@ -435,98 +403,147 @@ def get_and_write_background_peptide_and_kmer(peptide_set, kmer_dict, gene, ref_
+    :param spanning_index2: set. set of starting positions for which the kmer will cross junction 2
+    :param spanning_index1_2: set. set of starting positions for which the kmer will cross junction 1 and 2
+    :param translation_sorted_coord: np.array. exon coordinates modified by reading frame in translation order
+    :return: Coord namedtuples with:
+    start_v1 = int. start genomic coordinate in exon 1 of the kmer (may not be exon 1 of the peptide)
+    stop_v1 = int. stop genomic coordinate in exon 1 of the kmer (may not be exon 1 of the peptide)
+    start_v2 = int. start genomic coordinate in exon 2 of the kmer (may not be exon 2 of the peptide)
+    stop_v2 = int. stop genomic coordinate in exon 2 of the kmer (may not be exon 2 of the peptide)
+    start_v3 = int. start genomic coordinate in exon 3 of the kmer (may not be exon 3 of the peptide)
+    stop_v3 = int. stop genomic coordinate in exon 3 of the kmer (may not be exon 3 of the peptide)
+    '''
+
+    def shift_to_start(start_pos_kmer, strand, overhang=0, end_kmer=False):
+        if end_kmer:
+            shift = ((start_pos_kmer + k) * 3) + overhang
+        else:
+            shift = (start_pos_kmer * 3) + overhang
+        if strand == '-':
+            shift = -shift
+        return shift
 
+    start_E1_modi, stop_E1_modi, start_E2_modi, stop_E2_modi, start_E3_modi, stop_E3_modi = translation_sorted_coord
 
-def create_output_kmer(output_peptide, k, expr_list, kmer_database=None):
-    """Calculate the output kmer and the corresponding expression based on output peptide
+    # get the position(s) right after the junction(s)
+    if spanning_index1:
+        pos_after_jx1 = max(spanning_index1) + 1
+        first_cross_jx1 = min(spanning_index1)
+    else:
+        pos_after_jx1 = np.nan
+        first_cross_jx1 = np.nan
+    if spanning_index2:
+        pos_after_jx2 = max(spanning_index2) + 1
+    else:
+        pos_after_jx2 = np.nan
 
-    Parameters
-    ----------
-    output_peptide: OutputPeptide. Filtered output_peptide_list.
-    k: int. Specify k-mer length
-    expr_lists: List(Tuple). Filtered expr_list.
-    kmer_database: Set of kmers to be removed on the fly,
-    usually from a public database as uniprot. I and L equivalence is applied
+    # get the overhang
+    if (abs(stop_E1_modi - start_E1_modi) % 3): #TODO elegance code?
+        overhang1 = 3 - (abs(stop_E1_modi - start_E1_modi) % 3)
+    else:
+        overhang1 = 0
+    if ((abs(stop_E1_modi - start_E1_modi) + abs(stop_E2_modi - start_E2_modi)) % 3):
+        overhang2 = 3 - ((abs(stop_E1_modi - start_E1_modi) + abs(stop_E2_modi - start_E2_modi)) % 3)
+    else:
+        overhang2 = 0
+
+    # get the number of nucleotides from the kmer relative to the junction(s)
+    nt_kmer_left_jx1 = ((pos_after_jx1 - start_pos_kmer) * 3 ) - overhang1 #number of nucleotides before jx1 from the kmer
+    nt_kmer_right_jx1 = ((k - (pos_after_jx1 - start_pos_kmer)) * 3 ) + overhang1
```

Comment:
Why is the overhang added on both ends?

Reply:
nt_kmer_left_jx1 is the number of nucleotides from the kmer on the left side of the junction, nt_kmer_right_jx1 is the number of nucleotides from the kmer on the right side of the junction. We need an overhang because the junction can fall anywhere inside a codon and not just in-between. 
Shortly you can check that nt_kmer_left_jx1 + nt_kmer_right_jx1 = 3 * k with k = kmer length

Reply:
Yes this is correct the overhang is included, hence taking the junction coordinates is equivalent to taking the *_modi coordinates **on the junction site.** This is what I do here 
```
    # Crosses first junction only
    elif start_pos_kmer in spanning_index1:
        **stop_kmer_E1 = stop_E1_modi
        start_kmer_E2 = start_E2_modi**
```
Additionally I would like the start and end positions of the kmer. 
For this I look at the number of aminoacids ( reps. nucleotides) that end up on the left or right side of the junction. But this number does not include the little shift or overhang on the junction. Therefore I add or subtract it. 
```
    # Crosses first junction only
    elif start_pos_kmer in spanning_index1:
        **start_kmer_E1 = stop_E1_modi - nt_kmer_left_jx1**
        stop_kmer_E1 = stop_E1_modi
        start_kmer_E2 = start_E2_modi
        **stop_kmer_E2 = start_E2_modi + nt_kmer_right_jx1**
```